spacy/kb.pyx

```diff
@@ -0,0 +1,147 @@
+
+    property entity_id_:
+        """RETURNS (unicode): ID of this entity in the KB"""
+        def __get__(self):
+            return self.kb.vocab.strings[self.entity_id]
+
+    property entity_name:
+        """RETURNS (uint64): hash of the entity's KB name"""
+        def __get__(self):
+            entry_index = <int64_t>self.kb._entry_index.get(self.entity_id)
+            return self.kb._entries[entry_index].entity_name_hash
+
+    property entity_name_:
+        """RETURNS (unicode): name of this entity in the KB"""
+        def __get__(self):
+            return self.kb.vocab.strings[self.entity_name]
+
+    property alias:
+        """RETURNS (uint64): hash of the alias"""
+        def __get__(self):
+            return self.alias_hash
+
+    property alias_:
+        """RETURNS (unicode): ID of the original alias"""
+        def __get__(self):
+            return self.kb.vocab.strings[self.alias]
+
+    property prior_prob:
+        def __get__(self):
+            return self.prior_prob
+
+
+cdef class KnowledgeBase:
+
+    def __init__(self, Vocab vocab):
+        self.vocab = vocab
+        self._entry_index = PreshMap()
+        self._alias_index = PreshMap()
+        self.mem = Pool()
+        self._create_empty_vectors()
+
+    def __len__(self):
+        return self.get_size_entities()
+
+    def get_size_entities(self):
+        return self._entries.size() - 1  # not counting dummy element on index 0
+
+    def get_size_aliases(self):
+        return self._aliases_table.size() - 1 # not counting dummy element on index 0
+
+    def add_entity(self, unicode entity_id, unicode entity_name=None, float prob=0.5, vectors=None, features=None):
+        """
+        Add an entity to the KB.
+        Return the hash of the entity ID at the end
+        """
+        if not entity_name:
+            entity_name = entity_id
+        cdef hash_t id_hash = self.vocab.strings.add(entity_id)
+        cdef hash_t name_hash = self.vocab.strings.add(entity_name)
+
+        # Return if this entity was added before
+        if id_hash in self._entry_index:
+            user_warning("Entity " + entity_id + " already exists in the KB")
+            return
+
+        cdef int32_t dummy_value = 342
+        self.c_add_entity(entity_id_hash=id_hash, entity_name_hash=name_hash, prob=prob,
+                          vector_rows=&dummy_value, feats_row=dummy_value)
+        # TODO self._vectors_table.get_pointer(vectors),
+        # self._features_table.get(features))
+
+        return id_hash
+
+    def add_alias(self, unicode alias, entities, probabilities):
+        """
+        For a given alias, add its potential entities and prior probabilies to the KB.
+        Return the alias_hash at the end
+        """
+
+        # Throw an error if the length of entities and probabilities are not the same
+        if not len(entities) == len(probabilities):
+            raise ValueError("The vectors for entities and probabilities for alias '" + alias
```

Comment:
Not super important at this point, but these messages should probably all move to `spacy.errors` later on.